### PR TITLE
Reduce React bloat on equivalent objects with similar temporal alias trees

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

In an attempt to pull out some of the work from https://github.com/facebook/prepack/pull/2148 to make things easier to consume and review. This PR is one part, where we check the temporal alias values (if an object value has it) and see if we can dedupe the trees. Example below:

Test:
```js
function App(props) {
  function f() {}
  var foo3 = Object.assign({}, props);
  var a = <span {...foo3} key={null} />
  var b = <span {...foo3} key={null} />
  return [a, b]
}
```

Compiled output before:
```js
  var _0 = function (props, context) {
    var _F = _$I;

    var _$0 = _F({}, props);

    var _$1 = {
      key: null
    };

    var _$2 = _F({}, _$0, _$1);

    var _J = {
      key: null
    };
    var _$4 = _J;
    var _4 = {};

    var _$5 = _F(_4, _$2, _$4);

    var _$6 = {
      key: null
    };

    var _$7 = _F({}, _$0, _$6);

    var _$9 = _J;
    var _7 = {};

    var _$A = _F(_7, _$7, _$9);

    var _2 = <span {..._4} />;

    var _5 = <span {..._7} />;

    return [_2, _5];
  };
```

With this PR:
```js
  var _0 = function (props, context) {
    var _A = _$G;

    var _$0 = _A({}, props);

    var _$1 = {
      key: null
    };

    var _$2 = _A({}, _$0, _$1);

    var _4 = {};

    var _$5 = _A(_4, _$2, _$1);

    var _2 = <span {..._4} />;

    return [_2, _2];
  };
```

It drops the size of our internal bundle by around 100 lines of code compared to master.

I realize too that I'll need to refactor this PR with the new React test infra changes. I'll do that once this PR has been approved.

### CODE REVIEW

**Comment by sebmarkbage** [comment]
I'd like to get @NTillmann to review the temporalAliasArgs (perhaps re-review the original PR that added it). This seems like a possibly fragile approach since there is so much relying on temporal-ness of arguments that reordering things isn't safe. E.g. with regard to havocing and other things emitting temporal initialization.

**Comment by trueadm** [comment]
@sebmarkbage I'm not sure I understand the concerns around `temporalAliasArgs`. It's designed to provide us a way to traverse previous snapshots that come from Object.assign or React. The ordering doesn't matter either. All these objects should be final and havocing shouldn't be affected.

File: src/values/ObjectValue.js [hunk 1]
```diff
@@ -570,12 +570,17 @@ export default class ObjectValue extends ConcreteValue {
...
+      let result = AbstractValue.createTemporalFromBuildFunction(this.$Realm, ObjectValue, temporalArgs, ([x]) => x, {
         skipInvariant: true,
         isPure: true,
       });
       invariant(result instanceof AbstractObjectValue);
       result.values = new ValuesDomain(template);
+      // Store the args for the temporal so we can easily clone
+      // and reconstruct the temporal at another point, rather than
+      // mutate the existing temporal
+      realm.temporalAliasArgs.set(result, temporalArgs);
```
Comment (hermanventer) [inline src/values/ObjectValue.js]: Why use a map to get at the arguments that are already stored in result and that can be obtained with result.args?

Comment (trueadm) [reply src/values/ObjectValue.js]: They're not on the `AbstractValue`, rather they are the temporal generator node `args`. If the abstract value and the generator node both had the same args, this entire thing would be vastly easier to do :)

Comment (trueadm) [reply src/values/ObjectValue.js]: Another way would be to link the generator entry to the abstract value, in fact that might be a general easier approach. In the other PR https://github.com/facebook/prepack/pull/2148, which this is a small part of, we also need to know the additional args when the generator entry was created so we can clone and re-create the generator entries. So having a link from an abstract value, and it's original temporal entry would be highly beneficial. If not, and you feel this is a bad approach, how might we do this? @sebmarkbage @NTillmann

Comment (hermanventer) [reply src/values/ObjectValue.js]: OK, looking at the code of generator.deriveAbstract I can't off hand see a compelling reason why the arguments should not also be stored on the abstract value. Perhaps there is some subtle reason arising from the structure of the serializer, so @NTillmann would have to weigh in on this.

Comment (trueadm) [reply src/values/ObjectValue.js]: @hermanventer Good point. I've created a PR to change to that approach https://github.com/facebook/prepack/pull/2199. I tried to adds the args (to see what would happen) and it appears to break the serializer output and we run into invariants, so I guess the derived abstract values don't have args on the value for structure reasons like you suggest.

File: test/react/FirstRenderOnly/equivalence3.js [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
+  var foo = Object.assign({}, props, foo ? x : y);
   var a = <span {...foo} key={null} />;
   var b = <span {...foo} key={null} />;
   return [a, b];
 }
 
 App.getTrials = function(renderer, Root, data, isCompiled) {
   if (isCompiled) {
-    const [a, b] = App({});
+    const [a, b] = App({foo: false});
```
Comment (gaearon) [inline test/react/FirstRenderOnly/equivalence3.js]: Wait. This looks confusing.

We shouldn't ever be using `App` in this test directly I think. It always led to bugs in the past. It should be using `Root` instead.

Was this intentional?

**Pull request closed without merging**